In [1]:
import pandas as pd

import pprint
pp = pprint.PrettyPrinter(indent =4)

There are 9 csv files of available items that needs to be concatenate
I will first read those files and save in dictionaries

In [2]:
#Read the csv files into pandas dataframe

In [102]:
list_of_products_2000 = pd.read_csv("list_of_products_2000.csv") 
list_of_products_2000=list_of_products_2000.to_dict('records')

list_of_available_products_Rerun_2000 = pd.read_csv("list_of_available_products_Rerun_2000.csv") 
list_of_available_products_Rerun_2000=list_of_available_products_Rerun_2000.to_dict('records')

list_of_products_4000 = pd.read_csv("list_of_products_4000.csv") 
list_of_products_4000=list_of_products_4000.to_dict('records')

list_of_available_products_Rerun_4000 = pd.read_csv("list_of_available_products_Rerun_4000.csv") 
list_of_available_products_Rerun_4000=list_of_available_products_Rerun_4000.to_dict('records')

list_of_products_6000 = pd.read_csv("list_of_products_6000.csv") 
list_of_products_6000=list_of_products_6000.to_dict('records')

list_of_available_products_Rerun_6000 = pd.read_csv("list_of_available_products_Rerun_6000.csv") 
list_of_available_products_Rerun_6000=list_of_available_products_Rerun_6000.to_dict('records')

list_of_products_8000 = pd.read_csv("list_of_products_8000.csv") 
list_of_products_8000=list_of_products_8000.to_dict('records')

list_of_products_10000 = pd.read_csv("list_of_products_10000.csv") 
list_of_products_10000=list_of_products_10000.to_dict('records')

list_of_products_12000 = pd.read_csv("list_of_products_12000.csv") 
list_of_products_12000=list_of_products_12000.to_dict('records')

In [103]:
pp.pprint(list_of_products_2000[0])
pp.pprint(list_of_available_products_Rerun_2000[0])
pp.pprint(list_of_products_4000[:10])
pp.pprint(list_of_available_products_Rerun_4000[0])
pp.pprint(list_of_products_6000[:10])
pp.pprint(list_of_available_products_Rerun_6000[0])
pp.pprint(list_of_products_8000[0])
pp.pprint(list_of_products_10000[0])
pp.pprint(list_of_products_12000[0])

{   'No_of_Rating': '1,391 ratings',
    'Unnamed: 0': 0,
    'asin': '9788072216',
    'category': 'Fragrance',
    'description': 'Brand Prada Scent Honey, Musk , Vanilla Item Form Spray '
                   'Item Weight 278 Grams Item Volume 1.7 Fluid Ounces',
    'name': 'Prada Candy by Prada for Women 1.7 oz Eau de Parfum Spray',
    'price': '$56.29',
    'rating': '4.7 out of 5 stars'}
{   'No_of_Rating': '687 ratings',
    'Unnamed: 0': 0,
    'asin': 'B000052YQU',
    'category': 'Skin Care',
    'description': 'Special Ingredients\n'
                   '\n'
                   '\n'
                   'Keratin\n'
                   '\n'
                   '\n'
                   '\n'
                   '\n'
                   'Item Form\n'
                   '\n'
                   '\n'
                   'Cream\n'
                   '\n'
                   '\n'
                   '\n'
                   '\n'
                   'Brand\n'
                   '\n'
                

In [104]:
Final_list_of_available_products = list_of_products_2000 + list_of_available_products_Rerun_2000 + list_of_products_4000 + list_of_available_products_Rerun_4000 + list_of_products_6000 + list_of_available_products_Rerun_6000 + list_of_products_8000 + list_of_products_10000 + list_of_products_12000

In [105]:
len(Final_list_of_available_products)

4242

There are 4242 records scraped from amazon.com under the category health and beauty products
Next step is to clean the data and keep the information essential to analysis


# Data Cleaning

In [106]:
Final_list_of_available_products[:3]

[{'Unnamed: 0': 0,
  'asin': '9788072216',
  'name': 'Prada Candy by Prada for Women 1.7 oz Eau de Parfum Spray',
  'category': 'Fragrance',
  'description': 'Brand Prada Scent Honey, Musk , Vanilla Item Form Spray Item Weight 278 Grams Item Volume 1.7 Fluid Ounces',
  'price': '$56.29',
  'rating': '4.7 out of 5 stars',
  'No_of_Rating': '1,391 ratings'},
 {'Unnamed: 0': 1,
  'asin': 'B00004TMFE',
  'name': 'Avalon Organics Therapy Thickening Conditioner, Biotin B-Complex, 14 Oz',
  'category': 'Hair Care',
  'description': 'Brand Avalon Organics Scent Biotin B Hair Type Thin Liquid Volume 14 Fluid Ounces Ingredients Aloe Barbadensis Leaf Juice (Certified Organic Ingredient), Aqua (Water), Brassica Alcohol, Glycerin, Cocos Nucifera (Coconut) Oil (Certified Organic Ingredient), Brassicyl Isoleucinate Esylate, Gluconolactone, Capryloyl Glycerin/Sebacic Acid Copolymer, Cedrus Atlantica Wood Oil, Citrus Aurantifolia (Lime) Oil, Citrus Aurantium Dulcis (Orange) Peel Oil, Citrus Limon (Lemo

In [107]:
#Removing the column key Unnamed: 0
def new_dict(old_dict):
    n = old_dict.copy()
    n.pop('Unnamed: 0',None)
    return n

Final_list_of_available_products = list(map(new_dict,Final_list_of_available_products))

In [108]:
Final_list_of_available_products[:3]

[{'asin': '9788072216',
  'name': 'Prada Candy by Prada for Women 1.7 oz Eau de Parfum Spray',
  'category': 'Fragrance',
  'description': 'Brand Prada Scent Honey, Musk , Vanilla Item Form Spray Item Weight 278 Grams Item Volume 1.7 Fluid Ounces',
  'price': '$56.29',
  'rating': '4.7 out of 5 stars',
  'No_of_Rating': '1,391 ratings'},
 {'asin': 'B00004TMFE',
  'name': 'Avalon Organics Therapy Thickening Conditioner, Biotin B-Complex, 14 Oz',
  'category': 'Hair Care',
  'description': 'Brand Avalon Organics Scent Biotin B Hair Type Thin Liquid Volume 14 Fluid Ounces Ingredients Aloe Barbadensis Leaf Juice (Certified Organic Ingredient), Aqua (Water), Brassica Alcohol, Glycerin, Cocos Nucifera (Coconut) Oil (Certified Organic Ingredient), Brassicyl Isoleucinate Esylate, Gluconolactone, Capryloyl Glycerin/Sebacic Acid Copolymer, Cedrus Atlantica Wood Oil, Citrus Aurantifolia (Lime) Oil, Citrus Aurantium Dulcis (Orange) Peel Oil, Citrus Limon (Lemon) Peel Oil, Citrus Paradisi (Grapefru

In [109]:
#Removing $ from price,  out of 5 stars from ratings and  ratings from No_of_ratings
def cleandata(x):
    price = x['price']
    rating = x['rating']
    No_of_Rating = x['No_of_Rating']
    x['price'] = price.replace("$","")
    x['rating'] = rating.replace(" out of 5 stars","")
    x['No_of_Rating'] = No_of_Rating.replace(" ratings","")
    x['No_of_Rating']= x['No_of_Rating'].replace(",","")
    return x

In [110]:
cleandata({'asin': '9788072216',
  'name': 'Prada Candy by Prada for Women 1.7 oz Eau de Parfum Spray',
  'category': 'Fragrance',
  'description': 'Brand Prada Scent Honey, Musk , Vanilla Item Form Spray Item Weight 278 Grams Item Volume 1.7 Fluid Ounces',
  'price': '$56.29',
  'rating': '4.7 out of 5 stars',
  'No_of_Rating': '1,391 ratings'})

{'asin': '9788072216',
 'name': 'Prada Candy by Prada for Women 1.7 oz Eau de Parfum Spray',
 'category': 'Fragrance',
 'description': 'Brand Prada Scent Honey, Musk , Vanilla Item Form Spray Item Weight 278 Grams Item Volume 1.7 Fluid Ounces',
 'price': '56.29',
 'rating': '4.7',
 'No_of_Rating': '1391'}

In [111]:
Final_list_of_available_products = list(map(lambda x: cleandata(x), Final_list_of_available_products))

In [112]:
Final_list_of_available_products[:1]

[{'asin': '9788072216',
  'name': 'Prada Candy by Prada for Women 1.7 oz Eau de Parfum Spray',
  'category': 'Fragrance',
  'description': 'Brand Prada Scent Honey, Musk , Vanilla Item Form Spray Item Weight 278 Grams Item Volume 1.7 Fluid Ounces',
  'price': '56.29',
  'rating': '4.7',
  'No_of_Rating': '1391'}]

There are some values for key price that is of the format
8.4-29.99. The following function filters out all the dictionaries with the format and return their indices

In [113]:
Range_index = [index for (index, d) in enumerate(Final_list_of_available_products) if d['price'].find('-')>0]
Range_index

[1312, 1500]

In [101]:
Final_list_of_available_products[1312]

{'asin': 'B0015PGFQW',
 'name': 'TIGI Bed Head Manipulator, Texture Paste, 2 oz / 57 g',
 'category': 'Hair Care',
 'description': 'Scent\n\n\n2 PACK\n\n\n\n\nBrand\n\n\nTIGI\n\n\n\n\nItem Form\n\n\nClay',
 'price': '19.195',
 'rating': '4.6',
 'No_of_Rating': '61'}

There are two dictionaries with range of entries at index 1312 and 1500. I will modify the price to average
I have defined a function that takes the range and run through the list to calculate its average
Since list is mutable, it reassigns the average and to keep it even, I have converted the float to int

In [114]:
def modify_price(list, Range_index):
    for i in Range_index:
        dic = list[i]  
        a = dic['price']
        b = a.replace("-"," ")
        c = b.split()
        aver = (float(c[0])+float(c[1]))/2
        dic['price'] = str(aver)
    



In [115]:
modify_price(Final_list_of_available_products,Range_index)

In [117]:
Final_list_of_available_products[1500]

{'asin': 'B001CSNXLC',
 'name': '2 Month Supply',
 'category': 'Wall Art',
 'description': 'Brand\n\n\nKirkland Signature\n\n\n\n\nItem Weight\n\n\n0.15 Pounds\n\n\n\n\nAge Range (Description)\n\n\nAdult',
 'price': '42.375',
 'rating': '4.2',
 'No_of_Rating': '1067'}

In [118]:
#changing price and rating to float and No_of ratings to string
def strtonum(x):
    price = x['price']
    rating = x['rating']
    No_of_Rating = x['No_of_Rating']
    x['price'] = float(price)
    x['rating'] = float(rating)
    x['No_of_Rating'] = int(No_of_Rating)
    return x

In [119]:
Final_list_of_available_products = list(map(lambda x: strtonum(x), Final_list_of_available_products))

In [120]:
Final_list_of_available_products[:3]

[{'asin': '9788072216',
  'name': 'Prada Candy by Prada for Women 1.7 oz Eau de Parfum Spray',
  'category': 'Fragrance',
  'description': 'Brand Prada Scent Honey, Musk , Vanilla Item Form Spray Item Weight 278 Grams Item Volume 1.7 Fluid Ounces',
  'price': 56.29,
  'rating': 4.7,
  'No_of_Rating': 1391},
 {'asin': 'B00004TMFE',
  'name': 'Avalon Organics Therapy Thickening Conditioner, Biotin B-Complex, 14 Oz',
  'category': 'Hair Care',
  'description': 'Brand Avalon Organics Scent Biotin B Hair Type Thin Liquid Volume 14 Fluid Ounces Ingredients Aloe Barbadensis Leaf Juice (Certified Organic Ingredient), Aqua (Water), Brassica Alcohol, Glycerin, Cocos Nucifera (Coconut) Oil (Certified Organic Ingredient), Brassicyl Isoleucinate Esylate, Gluconolactone, Capryloyl Glycerin/Sebacic Acid Copolymer, Cedrus Atlantica Wood Oil, Citrus Aurantifolia (Lime) Oil, Citrus Aurantium Dulcis (Orange) Peel Oil, Citrus Limon (Lemon) Peel Oil, Citrus Paradisi (Grapefruit) Peel Oil, Eucalyptus Globul

Now I have cleaned and formated my scraped data. I will convert my list of dictionary to dataframe

In [121]:
Final_list_of_available_products_df = pd.DataFrame(Final_list_of_available_products)

In [122]:
Final_list_of_available_products_df.head()

asin                                               name   category  \
0  9788072216  Prada Candy by Prada for Women 1.7 oz Eau de P...  Fragrance   
1  B00004TMFE  Avalon Organics Therapy Thickening Conditioner...  Hair Care   
2  B00004TUBL  CLASSIC Better Living Two Chamber Dispenser, W...       Bath   
3  B00004TUBV  Better Living Products 73350 ULTI-MATE 3 Chamb...       Bath   
4  B000052WYD  Maybelline New York Cover Stick Concealer, Lig...     Makeup   

                                         description  price  rating  \
0  Brand Prada Scent Honey, Musk , Vanilla Item F...  56.29     4.7   
1  Brand Avalon Organics Scent Biotin B Hair Type...   7.81     4.1   
2  Color White Brand CLASSIC Item Dimensions LxWx...  39.99     4.3   
3  Color White Material Plastic Brand Better Livi...  38.21     4.5   
4  Color Ivory Brand Maybelline New York Ingredie...   4.99     4.6   

   No_of_Rating  
0          1391  
1          5646  
2            99  
3          1245  
4          3316

In [224]:
type(Final_list_of_available_products_df)

pandas.core.frame.DataFrame

In [225]:
group_category = Final_list_of_available_products_df.groupby('category')

In [226]:
group_category.size()

category
Baby & Child Care                         2
Baby Care                                11
Bath                                      6
Bedding                                   4
Beverages                                 2
Camera & Photo                            1
Cases, Holsters & Sleeves                 1
Costumes & Accessories                    4
Diapering                                 1
Foot, Hand & Nail Care                  278
Fragrance                               268
Furniture                                 1
Hair Care                              1228
Health Care                              51
Home Décor Products                       2
Household Supplies                        7
Kitchen & Dining                          3
Luggage & Travel Gear                     1
Makeup                                  712
Medical Supplies & Equipment              3
Men                                       1
Oral Care                                 2
Outdoor Recreation     

In [123]:
#Writing into csv file
Final_list_of_available_products_df.to_csv('List_of_final_wrangled_product.csv')

I will compare the columns in old dataset and columns in scraped dataset and keep the columns required for my analysis 

In [172]:
Review = pd.read_json('reviews_Beauty_5.json',lines = True)
Review.head(10)

reviewerID        asin                    reviewerName helpful  \
0  A1YJEY40YUW4SE  7806397051                          Andrea  [3, 4]   
1   A60XNB876KYML  7806397051                      Jessica H.  [1, 1]   
2  A3G6XNM240RMWA  7806397051                           Karen  [0, 1]   
3  A1PQFP6SAJ6D80  7806397051                           Norah  [2, 2]   
4  A38FVHZTNQ271F  7806397051                       Nova Amor  [0, 0]   
5  A3BTN14HIZET6Z  7806397051  S. M. Randall "WildHorseWoman"  [1, 2]   
6  A1Z59RFKN0M5QL  7806397051               tasha "luvely12b"  [1, 3]   
7   AWUO9P6PL1SY8  7806397051                   TreMagnifique  [0, 1]   
8  A3LMILRM9OC3SA  9759091062                             NaN  [0, 0]   
9  A30IP88QK3YUIO  9759091062             Amina Bint Ibraheem  [0, 0]   

                                          reviewText  overall  \
0  Very oily and creamy. Not at all what I expect...        1   
1  This palette was a decent price and I was look...        3   
2  The texture of this concealer pallet is fantas...        4   
3  I really can't tell what exactly this thing is...        2   
4  It was a little smaller than I expected, but t...        3   
5  I was very happy to get this palette, now I wi...        5   
6  PLEASE DONT DO IT! this just rachett the palet...        1   
7  Chalky,Not Pigmented,Wears off easily,Not a Co...        2   
8  Did nothing for me. Stings when I put it on. I...        2   
9  I bought this product to get rid of the dark s...        3   

                                             summary  unixReviewTime  \
0                             Don't waste your money      1391040000   
1                                        OK Palette!      1397779200   
2                                      great quality      1378425600   
3                             Do not work on my face      1386460800   
4                                         It's okay.      1382140800   
5                                 Very nice palette!      1365984000   
6                                             smh!!!      1376611200   
7  Chalky, Not Pigmented, Wears off easily, Not a...      1378252800   
8        no Lightening, no Brightening,......NOTHING      1405209600   
9                                        Its alright      1388102400   

    reviewTime  
0  01 30, 2014  
1  04 18, 2014  
2   09 6, 2013  
3   12 8, 2013  
4  10 19, 2013  
5  04 15, 2013  
6  08 16, 2013  
7   09 4, 2013  
8  07 13, 2014  
9  12 27, 2013

In [173]:
print("These are the columns in old dataset")
print(Review.columns)
print("These are the columns in scraped dataset")
print(Final_list_of_available_products_df.columns)

These are the columns in old dataset
Index(['reviewerID', 'asin', 'reviewerName', 'helpful', 'reviewText',
       'overall', 'summary', 'unixReviewTime', 'reviewTime'],
      dtype='object')
These are the columns in scraped dataset
Index(['asin', 'name', 'category', 'description', 'price', 'rating',
       'No_of_Rating'],
      dtype='object')


1. asin id: This is the primary key for my scraped data and will be used to join the two dataframes
2. reviewer id: I will keep this for future plan of building a recommender system
3. helpful: This column will help in sentimental analysis 
4. overall: This were the ratings in 2014 and will be averaged to compare with the rating in 2021 which is scraped to see the difference to find a correlation with sentiment analysis
5. Name and description: The scraped name and description of the product, I will keep it for my NLP and sentiment analysis
6. Category: This is the scraped data and will be helpful in analytics
7. price: This will be kept for price analysis
8. Rating: This is the 2021 rating and will be used to see the change from 2014
9. No_of_Ratings: This gives an idea about the number of people who purchased the product
10. reviewText: This is the review from 2014 and will be used in NLP for sentiment analysis
11. Summary : Short heading of the review, again helpful in NLP

I will drop columns, reviewerName, unixReviewTime and reviewtime from old dataset

In [174]:
Review = Review.drop(['reviewerName', 'unixReviewTime', 'reviewTime'], axis=1)

In [175]:
Review.head()

reviewerID        asin helpful  \
0  A1YJEY40YUW4SE  7806397051  [3, 4]   
1   A60XNB876KYML  7806397051  [1, 1]   
2  A3G6XNM240RMWA  7806397051  [0, 1]   
3  A1PQFP6SAJ6D80  7806397051  [2, 2]   
4  A38FVHZTNQ271F  7806397051  [0, 0]   

                                          reviewText  overall  \
0  Very oily and creamy. Not at all what I expect...        1   
1  This palette was a decent price and I was look...        3   
2  The texture of this concealer pallet is fantas...        4   
3  I really can't tell what exactly this thing is...        2   
4  It was a little smaller than I expected, but t...        3   

                  summary  
0  Don't waste your money  
1             OK Palette!  
2           great quality  
3  Do not work on my face  
4              It's okay.

In [176]:
Review[['helpful1','Not helpful']] = pd.DataFrame(Review.helpful.tolist(), index= Review.index)

In [177]:
Review.head()

reviewerID        asin helpful  \
0  A1YJEY40YUW4SE  7806397051  [3, 4]   
1   A60XNB876KYML  7806397051  [1, 1]   
2  A3G6XNM240RMWA  7806397051  [0, 1]   
3  A1PQFP6SAJ6D80  7806397051  [2, 2]   
4  A38FVHZTNQ271F  7806397051  [0, 0]   

                                          reviewText  overall  \
0  Very oily and creamy. Not at all what I expect...        1   
1  This palette was a decent price and I was look...        3   
2  The texture of this concealer pallet is fantas...        4   
3  I really can't tell what exactly this thing is...        2   
4  It was a little smaller than I expected, but t...        3   

                  summary  helpful1  Not helpful  
0  Don't waste your money         3            4  
1             OK Palette!         1            1  
2           great quality         0            1  
3  Do not work on my face         2            2  
4              It's okay.         0            0

In [178]:
#Each asin has multiple reviewer, I will first concatenate the reviews and average
#out the rating and store into a new data frame
Review.dtypes

reviewerID     object
asin           object
helpful        object
reviewText     object
overall         int64
summary        object
helpful1        int64
Not helpful     int64
dtype: object

In [179]:
#Dropping the row helpful
Review = Review.drop('helpful', axis=1)

In [210]:
Review.head()

reviewerID        asin  \
0  A1YJEY40YUW4SE  7806397051   
1   A60XNB876KYML  7806397051   
2  A3G6XNM240RMWA  7806397051   
3  A1PQFP6SAJ6D80  7806397051   
4  A38FVHZTNQ271F  7806397051   

                                          reviewText  overall  \
0  Very oily and creamy. Not at all what I expect...        1   
1  This palette was a decent price and I was look...        3   
2  The texture of this concealer pallet is fantas...        4   
3  I really can't tell what exactly this thing is...        2   
4  It was a little smaller than I expected, but t...        3   

                  summary  helpful1  Not helpful  
0  Don't waste your money         3            4  
1             OK Palette!         1            1  
2           great quality         0            1  
3  Do not work on my face         2            2  
4              It's okay.         0            0

In [194]:
group_asin = Review.groupby('asin')

In [195]:
avg_rating = group_asin.overall.mean().reset_index(name = "avg_rating")

In [196]:
avg_rating.head()

asin  avg_rating
0  7806397051    2.625000
1  9759091062    3.090909
2  9788072216    5.000000
3  9790790961    4.333333
4  9790794231    3.600000

In [200]:
#Writing the dataframe into a csv file
avg_rating.to_csv('avg_rating.csv')

In [203]:
avg_rating.shape

(12101, 2)

In [215]:
Review_concat = group_asin.reviewText.apply(' '.join).reset_index(name = "review_concat")

In [216]:
type(Review_concat)

pandas.core.frame.DataFrame

In [217]:
Review_concat.head()

asin                                      review_concat
0  7806397051  Very oily and creamy. Not at all what I expect...
1  9759091062  Did nothing for me. Stings when I put it on. I...
2  9788072216  I haven't been a big fan of Prada's fragrances...
3  9790790961  I'm very picky when it comes to fragrance. I l...
4  9790794231  I like this perfume. I first discovered it whe...

In [220]:
Review_concat.iloc[0,1]

"Very oily and creamy. Not at all what I expected... ordered this to try to highlight and contour and it just looked awful!!! Plus, took FOREVER to arrive. This palette was a decent price and I was looking for a few different shades. This palette conceals decently, however, it does somewhat cake up and crease. The texture of this concealer pallet is fantastic, it has great coverage and a wide variety of uses, I guess it's meant for professional makeup artists and a lot of the colours are of no use to me but I use at least two of them on a regular basis, and two more occasionally, which is the only reason I'm giving it for stars, I feel like the range of colors is kind of a waste for me, but the  product itself  is wonderful, it's not cakey, gives me a natural for and concealed my imperfections, therefore I highly recommend it :) I really can't tell what exactly this thing is. It's not powder but a kind of oil-ish pasty fluid. And so far I tried twice but it doesnt really show any color

In [221]:
#Writing the dataframe to csv file
Review_concat.to_csv('Review_concat.csv')

I will analyze the following dataframes:
1. average rating for every unique asin in 2014
2. average rating for every unique asin in 2021
3. Sentiment Analysis of Reviews in 2014

Next steps:
1. Exploratory analysis of old and scraped data
2. Find the change in average ratings from 2014 to 2021 (delta)
3. List down the top 10 positive and negative delta from categories (skincare, makeup and haircare since these three categories have most number of scraped data)
4. Scrape the present reviews of those 20 products from these three categories and use NPL for sentiment analysis